execution_engine/src/runtime/mod.rs

```diff
@@ -1759,6 +1760,18 @@ where
             return Err(ExecError::LockedEntity(package_hash));
         }
 
+        // WHAT HAPPENED TO THE PERMISSION CHECK TO MAKE SURE THE CALLER CAN UPGRADE THIS CONTRACT?
+        //
+        // STEP 1: LOAD THE CONTRACT AND CHECK IF CALLER IS IN ASSOCIATED KEYS WITH ENOUGH WEIGHT
+        //     TO UPGRADE (COMPARE TO THE ACTION THRESHOLD FOR UPGRADE ACTION).
+        // STEP 2: IF CALLER IS NOT IN CONTRACTS ASSOCIATED KEYS
+        //    CHECK FOR LEGACY UREFADDR UNDER KEY:HASH(PACKAGEADDR)
+        //    IF FOUND,
+        //      call validate_uref(that uref)
+        //    IF VALID,
+        //      create the new contract version carrying forward previous state including associated
+        // keys      BUT add the caller to the associated keys with weight == to the action
+        // threshold for upgrade ELSE, error
```

Comment:
move this down into the later function where the action actually occurs, and massage the language if necessary

Reply:
Will do!